pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,203 @@
+
+    return qml.registers(temp_register_dict)
+
+
+class FromBloq(Operation):
+    r"""
+    A shim for using bloqs as a PennyLane :class:`~.Operation`.
+
+    Args:
+        bloq (Bloq): the bloq to wrap
+        wires (WiresLike): the wires to act on
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    from qualtran.bloqs.basic_gates import CNOT
+
+    >>> qualtran_toffoli = qml.FromBloq(CNOT(), [0, 1])
+    >>> qualtran_toffoli.matrix()
+    array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+       [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j]])
+
+    A simple example showcasing how to use `qml.FromBloq` inside a device:
+
+    .. code-block::
+
+        dev = qml.device("default.qubit")
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.FromBloq(XGate(), [0])
+            return qml.expval(qml.Z(wires=[0]))
+
+    >>> circuit()
+    -1.0
+    """
+
+    def __init__(self, bloq, wires: WiresLike):
+        assert isinstance(bloq, Bloq)
+        self._hyperparameters = {"bloq": bloq}
+        super().__init__(wires=wires, id=None)
+
+    def __repr__(self):  # pylint: disable=protected-access
+        return f'FromBloq({self._hyperparameters["bloq"]}, wires={self.wires})'
+
+    def compute_decomposition(
+        self, wires, **kwargs
+    ):  # pylint: disable=arguments-differ, unused-argument
+        ops = []
+        bloq = self._hyperparameters["bloq"]
+
+        try:
+            # Bloqs need to be decomposed in order to access the connections
+            cbloq = bloq.decompose_bloq() if not isinstance(bloq, CompositeBloq) else bloq
+            temp_registers = _get_named_registers(cbloq.signature.lefts())
+            soq_to_wires = {
+                Soquet(LeftDangle, idx=idx, reg=reg): list(temp_registers[reg.name])
+                for reg in cbloq.signature.lefts()
+                for idx in reg.all_idxs()
+            }
+
+            for binst, pred_cxns, succ_cxns in cbloq.iter_bloqnections():
+                in_quregs = {
+                    reg.name: np.empty((*reg.shape, reg.bitsize), dtype=object).flatten()
+                    for reg in binst.bloq.signature.lefts()
+                }
+                # The out_quregs inform us of the total # of wires in the circuit to account for
+                # wires that are split or allocated in the cbloq
+                out_quregs = {
+                    reg.name: np.empty((*reg.shape, reg.bitsize), dtype=object).flatten()
+                    for reg in binst.bloq.signature.rights()
+                }
+
+                # This is to track the number of wires defined at the LeftDangle stage
+                # so if we need to add more wires, we know what index to start at
+                soq_to_wires_len = 0
+                if len(soq_to_wires.values()) > 0:
+                    try:
+                        soq_to_wires_len = list(soq_to_wires.values())[-1][-1] + 1
```

Comment:
We can take out `list(soq_to_wires.values())[-1]` outside the try and except, and then use `instance` to access further deeper index, if possible.